src/Gui/NotificationBox.cpp

```diff
@@ -0,0 +1,352 @@
+    void placeNotificationLabel(const QPoint &pos);
+
+    /// The instance
+    static NotificationLabel *instance;
+protected:
+    void paintEvent(QPaintEvent *e) override;
+    void resizeEvent(QResizeEvent *e) override;
+private:
+    /// Re-start the notification expiration timer
+    void restartExpireTimer(int msecDisplayTime);
+    /// Hide notification right away
+    void hideNotificationImmediately();
+
+private:
+    int minShowTime;
+    QTimer hideTimer;
+    QTimer expireTimer;
+};
+
+NotificationLabel *NotificationLabel::instance = nullptr;
+
+NotificationLabel::NotificationLabel(const QString &text, const QPoint &pos, QWidget *w, int msecDisplayTime, int minShowTime)
+: QLabel(w, Qt::ToolTip | Qt::BypassGraphicsProxyWidget), minShowTime(minShowTime)
+{
+    delete instance;
+    instance = this;
+    setForegroundRole(QPalette::ToolTipText); // defaults to ToolTip QPalette
+    setBackgroundRole(QPalette::ToolTipBase); // defaults to ToolTip QPalette
+    setPalette(NotificationBox::palette());
+    ensurePolished();
+    setMargin(1 + style()->pixelMetric(QStyle::PM_ToolTipLabelFrameWidth, nullptr, this));
+    setFrameStyle(QFrame::NoFrame);
+    setAlignment(Qt::AlignLeft);
+    setIndent(1);
+    qApp->installEventFilter(this);
+    setWindowOpacity(style()->styleHint(QStyle::SH_ToolTipLabel_Opacity, nullptr, this) / 255.0);
+    setMouseTracking(false);
+    hideTimer.setSingleShot(true);
+    expireTimer.setSingleShot(true);
+
+    expireTimer.callOnTimeout([this](){
+        hideTimer.stop();
+        expireTimer.stop();
+        hideNotificationImmediately();
+    });
+
+    hideTimer.callOnTimeout([this](){
+        expireTimer.stop();
+        hideNotificationImmediately();
+    });
+
+    reuseNotification(text, msecDisplayTime, pos);
+}
+void NotificationLabel::restartExpireTimer(int msecDisplayTime)
+{
+    int time = 10000 + 40 * qMax(0, text().length()-100);
+    if (msecDisplayTime > 0) {
+        time = msecDisplayTime;
+    }
+    expireTimer.start(time);
+    hideTimer.stop();
+}
+void NotificationLabel::reuseNotification(const QString &text, int msecDisplayTime, const QPoint &pos)
+{
+    setText(text);
+    updateSize(pos);
+    restartExpireTimer(msecDisplayTime);
+}
+void NotificationLabel::updateSize(const QPoint &pos)
+{
+    // Ensure that we get correct sizeHints by placing this window on the right screen.
+    QFontMetrics fm(font());
+    QSize extra(1, 0);
+    // Make it look good with the default ToolTip font on Mac, which has a small descent.
+    if (fm.descent() == 2 && fm.ascent() >= 11) {
+        ++extra.rheight();
+    }
+
+    setWordWrap(Qt::mightBeRichText(text()));
+
+    QSize sh = sizeHint();
+
+    // ### When the above WinRT code is fixed, windowhandle should be used to find the screen.
+    QScreen *screen = QGuiApplication::screenAt(pos);
+    if (!screen) {
+        screen = QGuiApplication::primaryScreen();
+    }
+
+    if (screen) {
```

Comment:
Yes.

Reply:
There is now an "else" as requested but this doesn't make much sense because the `if (!screen)` leads to dead code.

The current code is:
```
    QScreen *screen = QGuiApplication::screenAt(pos);
    if (!screen) {
/* If screen is null then the primary screen is accessed but then nothing further happens with it */
        screen = QGuiApplication::primaryScreen();
    }
    else {
        const qreal screenWidth = screen->geometry().width();
        if (!wordWrap() && sh.width() > screenWidth) {
            setWordWrap(true);
            sh = sizeHint();
        }
    }

```

Instead the layout should be
```
    QScreen *screen = QGuiApplication::screenAt(pos); // screen may become null
    if (!screen) {
        screen = QGuiApplication::primaryScreen();           // if null access the primary screen
    }
    if (screen) {                                                                        // it's not null now
        const qreal screenWidth = screen->geometry().width();
        if (!wordWrap() && sh.width() > screenWidth) {
            setWordWrap(true);
            sh = sizeHint();
        }
    }

```

Reply:
Good one. Obviously I did not pay enough attention when applying the requested change. Thanks.

I will review the patches for windows and this when I move to the next stage of merging NotificationArea. Then it will also be easier to do the testing.

Reply:
> Obviously I did not pay enough attention when applying the requested change

I did the original mistake when reviewing so I'm the one to blame. :D

Reply:
A maintainer is no better than a developer. Handling the review is a shared responsibility. It is for the maintainer to raise concerns and for the developer to evaluate the appropriateness of the concerns. That way the repository gets the code it deserves and we all learn in the process. For example, I learnt that the code review shall be seen in the full code window. If I had done this, it is like I would have realised. I have learnt ;)

But of course, we can share the blame if you insist LoL.